luaui/Widgets/cmd_area_commands_filter.lua

```diff
@@ -0,0 +1,489 @@
+	[GameCMD.UNIT_SET_TARGET] = { handle = defaultHandler, allowedTargetTypes = { [UNIT] = true } },
+	[GameCMD.UNIT_SET_TARGET_NO_GROUND] = { handle = defaultHandler, allowedTargetTypes = { [UNIT] = true } },
+	[CMD.RESURRECT] = { handle = defaultHandler, allowedTargetTypes = { [FEATURE] = true } },
+	[CMD.LOAD_UNITS] = { handle = loadUnitsHandler, allowedTargetTypes = { [UNIT] = true } },
+}
+
+local function filterUnits(targetId, cmdX, cmdZ, radius, options, skipAlliedUnits)
+	local filteredTargets = {}
+	local unitDefId = spGetUnitDefID(targetId)
+	if not unitDefId then
+		return nil
+	end
+
+	local isEnemyTarget = (spGetUnitAllyTeam(targetId) ~= myAllyTeamID)
+
+	local unitsInArea
+	if isEnemyTarget then
+		unitsInArea = spGetUnitsInCylinder(cmdX, cmdZ, radius, Spring.ENEMY_UNITS)
+	elseif not skipAlliedUnits then
+		unitsInArea = spGetUnitsInCylinder(cmdX, cmdZ, radius, myTeamID)
+	end
+	if not unitsInArea then
+		return nil
+	end
+
+	for i = 1, #unitsInArea do
+		local unitID = unitsInArea[i]
+		if options.ctrl or (options.alt and spGetUnitDefID(unitID) == unitDefId) then
+			table.insert(filteredTargets, unitID)
+		end
+	end
+	return filteredTargets
+end
+
+local function filterFeatures(targetId, cmdX, cmdZ, radius, options)
+	local filteredTargets = {}
+	local featureDefId = spGetFeatureDefID(targetId)
+	if not featureDefId then
+		return nil
+	end
+
+	local featuresInArea = spGetFeaturesInCylinder(cmdX, cmdZ, radius)
+	if not featuresInArea then
+		return nil
+	end
+
+	for i = 1, #featuresInArea do
+		local featureId = featuresInArea[i]
+		if options.alt and spGetFeatureDefID(featureId) == featureDefId then
+			-- featureId is normalised to Game.maxUnits + featureId because of:
+			-- https://springrts.com/wiki/Lua_CMDs#CMDTYPE.ICON_UNIT_FEATURE_OR_AREA
+			-- "expect 1 parameter in return (unitid or Game.maxUnits+featureid)"
+			featureId = Game.maxUnits + featureId
+			table.insert(filteredTargets, featureId)
+		end
+	end
+	return filteredTargets
+end
+
+function widget:CommandNotify(cmdId, params, options)
+	if not (options.alt or options.ctrl) then
+		return false
+	end
+
+	if #params ~= 4 then
+		return false
+	end
+
+	local currentCommand = allowedCommands[cmdId]
+	if not currentCommand then
+		return false
+	end
+
+	local selectedUnits = spGetSelectedUnits()
+	if #selectedUnits == 0 then
+		return false
+	end
+
+	local cmdX, cmdY, cmdZ, radius = params[1], params[2], params[3], params[4]
+	local mouseX, mouseY = spWorldToScreenCoords(cmdX, cmdY, cmdZ)
+	local targetType, targetId = spTraceScreenRay(mouseX, mouseY)
+
+	if not currentCommand.allowedTargetTypes[targetType] then
+		return false
+	end
+
+	local filteredTargets
+
+	if targetType == UNIT then
+		filteredTargets = filterUnits(targetId, cmdX, cmdZ, radius, options, currentCommand.skipAlliedUnits)
+	elseif targetType == FEATURE then
+		local featureDefName = spGetFeatureResurrect(targetId)
+		-- filter only wrecks which can be resurrected
```

Comment:
what's the benefit of disallowing filtering by non-rezzable wrecks? e.g. what if i want to specifically reclaim trees because my windgens were raided?